lib/Parse/Lexer.cpp

```diff
@@ -932,20 +932,27 @@ void Lexer::lexHexNumber() {
   const char *TokStart = CurPtr-1;
   assert(*TokStart == '0' && "not a hex literal");
 
-  // 0x[0-9a-fA-F][0-9a-fA-F_]*
-  ++CurPtr;
-  if (!isHexDigit(*CurPtr)) {
-    diagnose(CurPtr, diag::lex_expected_digit_in_int_literal);
+  auto expected_hex_digit = [&]() {
+    diagnose(CurPtr, diag::lex_expected_digit_in_int_literal, 3 /*Hex*/);
```

Comment:
The code might be slightly neater/more self-documenting with an `enum` like,

```cpp
enum class ExpectedDigitKind : unsigned {
    Binary,
    Octal, 
    Decimal,
    Hex
}
```

which would mean these calls could look like `diagnose(..., (unsigned)ExpectedDigitKind::Hex)` here and `expected_int_digit(..., ExpectedDigitKind::Binary)` below (after adjusting the closure to take `ExpectedDigitKind`).

(If you're interested in an example to crib off, the `where_without_generic_params` diagnostic does something similar with the `Parser::WhereClauseKind` enum.)